File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-14.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        //set = xy
        //zxyzxyz
        // l
        //   r  

        if(s.length() == 1) return 1;

        Set<Character> seen = new HashSet<>();
        int left = 0;
        int right = 0;
        int result = 0;

        while(right <= s.length() - 1){
            char current = s.charAt(right);

            while(seen.contains(current)){
                char leftValue = s.charAt(left);
                seen.remove(leftValue);
                left++;
            }

            result = Math.max(result, right - left + 1);

            seen.add(current);
            right++;
        }

       //result = Math.max(result, right - left + 1);

        return result;
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-14.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left = 0
        right = len(nums) - 1
        result = math.inf

        while left <= right:

            middle = (left + right) // 2

            if nums[left] < nums[middle] and nums[middle] < nums[right]:
                result = min(nums[left], result)
            
            result = min(result, nums[middle])

            if nums[left] <= nums[middle]:
                left = middle + 1
            else:
                right = middle -1
    

        return result
            

